File: rede-bayseana-smile/tutorials/R/tutorial4.R
library(rSMILE)
source("License.R")

# Tutorial4 loads the XDSL file file created by Tutorial1
# and adds decision and utility nodes, which transforms 
# a Bayesian Network (BN) into an Influence Diagram (ID).

createNode = function(net, nodeType, id, name, outcomes, xPos, yPos) {
    handle <- net$addNode(nodeType, id)
    initialOutcomeCount <- net$getOutcomeCount(handle)
    count <- length(outcomes)
    net$setNodeName(handle, name)
    net$setNodePosition(handle, xPos, yPos, 85L, 55L)
    if (!is.null(outcomes)) {
        sapply(0:(initialOutcomeCount-1), 
            function(x) net$setOutcomeId(handle, x, outcomes[x+1]))
        if (initialOutcomeCount < count) {
            sapply(initialOutcomeCount:(count-1), 
                function(x) net$addOutcome(handle, outcomes[x+1]))
        }
    }

    return(handle)
}

cat("Starting Tutorial4...\n")
net <- Network()

net$readFile("tutorial1.xdsl")

s <- net$getNode("Success")

i <- createNode(net, net$NodeType$DECISION, "Invest", "Investment decision", 
    c("Invest", "DoNotInvest"), 160L, 240L)

g <- createNode(net, net$NodeType$UTILITY, "Gain", "Financial Gain", NULL, 60, 200)

net$addArc(i, g)
net$addArc(s, g)

gainDefinition <- c(10000, # Utility(Invest=I, Success=S)
                    -5000, # Utility(Invest=I, Success=F)
                    500,   # Utility(Invest=D, Success=S)
                    500)   # Utility(Invest=D, Success=F)


net$setNodeDefinition(g, gainDefinition)

net$writeFile("tutorial4.xdsl")                   

cat("Tutorial4 complete: Influence diagram written to tutorial4.xdsl.\n")

File: rede-bayseana-smile/tutorials/cpp/tutorial4.cpp
// tutorial4.cpp
// Tutorial4 loads the XDSL file file created by Tutorial1
// and adds decision and utility nodes, which transforms 
// a Bayesian Network (BN) into an Influence Diagram (ID).

#include "/home/josegustavo/Documentos/IC/rede-bayseana/include/smile.h"
#include "/home/josegustavo/Documentos/IC/rede-bayseana/include/smile_license.h"

#include <cstdio>


static int CreateNode(
    DSL_network& net, int nodeType, const char* id, const char* name,
    std::initializer_list<const char*> outcomes, int xPos, int yPos);


int Tutorial4()
{
    printf("Starting Tutorial4...\n");

    DSL_errorH().RedirectToFile(stdout);

    DSL_network net;
    // load the network created by Tutorial1
    int res = net.ReadFile("tutorial1.xdsl");
    if (DSL_OKAY != res)
    {
        printf(
            "Network load failed, did you run Tutorial1 before Tutorial4?\n");
        return res;
    }

    int s = net.FindNode("Success");
    if (s < 0)
    {
        printf("Success node not found.");
        return s;
    }

    int i = CreateNode(net, DSL_LIST, "Invest", "Investment decision", 
        { "Invest", "DoNotInvest" }, 160, 240);

    int g = CreateNode(net, DSL_TABLE, "Gain", "Financial gain", 
        {}, 60, 200);

    net.AddArc(i, g);
    net.AddArc(s, g);

    res = net.GetNode(g)->Def()->SetDefinition({10000, -5000, 500, 500});
    res = net.WriteFile("tutorial4.xdsl");
    if (DSL_OKAY != res)
    {
        return res;
    }

    printf("Tutorial4 complete: Influence diagram written to tutorial4.xdsl\n");
    return DSL_OKAY;
}


static int CreateNode(
    DSL_network &net, int nodeType, const char *id, const char *name, 
    std::initializer_list<const char*> outcomes, int xPos, int yPos)
{
    int handle = net.AddNode(nodeType, id);
    DSL_node *node = net.GetNode(handle);
    node->SetName(name);
    if (outcomes.size() > 0)
    {
        node->Def()->SetNumberOfOutcomes(outcomes);
    }
    DSL_rectangle &position = node->Info().Screen().position;
    position.center_X = xPos;
    position.center_Y = yPos;
    position.width = 85;
    position.height = 55;
    return handle;
}
